query_server/query/src/sql/planner.rs

```diff
@@ -1794,33 +1815,59 @@ impl<'a, S: ContextProviderExtension + Send + Sync + 'a> SqlPlanner<'a, S> {
         &self,
         stmt: ast::AlterUser,
         user: &User,
+        must_change: bool,
     ) -> Result<PlanWithPrivileges> {
         let ast::AlterUser { name, operation } = stmt;
 
-        let user_name = normalize_ident(name);
+        let sql_user_name = normalize_ident(name);
         // 查询用户信息，不存在直接报错咯
         // fn user(
         //     &self,
         //     name: &str
         // ) -> Result<Option<UserDesc>>;
-        let user_desc = self.schema_provider.get_user(&user_name).await?;
-        let user_id = *user_desc.id();
+        let sql_user_desc = self.schema_provider.get_user(&sql_user_name).await?;
+        let sql_user_id = *sql_user_desc.id();
 
-        let mut privileges = vec![Privilege::Global(GlobalPrivilege::User(Some(user_id)))];
+        let mut privileges = vec![Privilege::Global(GlobalPrivilege::User(Some(sql_user_id)))];
 
         let alter_user_action = match operation {
             AlterUserOperation::RenameTo(new_name) => {
                 AlterUserAction::RenameTo(normalize_ident(new_name))
             }
             AlterUserOperation::Set(sql_option) => {
-                let user_options = sql_options_to_user_options(vec![sql_option])?;
-                if user_desc.is_root_admin() && !user.desc().is_root_admin() {
+                let (mut sql_user_option, password) =
+                    sql_options_to_user_options(vec![sql_option])?;
+                let user_desc = user.desc();
+                if sql_user_option.must_change_password().is_some() {
+                    privileges = vec![Privilege::Global(GlobalPrivilege::System)];
+                }
+
+                if must_change {
+                    if *user_desc.id() == sql_user_id && sql_user_option.hash_password().is_some() {
+                        sql_user_option.change_password();
+                    } else {
+                        return Err(QueryError::InsufficientPrivileges {
+                            privilege: "change password".to_string(),
+                        });
+                    }
+                }
+                if let Some(hash) = user_desc.options().hash_password() {
+                    if *user_desc.id() == sql_user_id
```

Comment:
I don't think we need a "same user" check, because user1 can't change user2's password to be the same as the old one